align_system/utils/alignment_utils.py

```diff
@@ -80,11 +85,24 @@ def __call__(self, kdma_values, target_kdmas, misaligned=False):
 
         selected_choice, probs = self._select_min_dist_choice(choices, distances, misaligned)
 
-        return selected_choice, probs
+        if return_best_sample_index:
+            best_sample_index = 0
+            min_dist = float('inf')
+            for sample_idx in range(len(samples)):
+                dist = 0
+                for target_kdma in target_kdmas:
+                    sample = kdma_values[selected_choice][target_kdma['kdma']][sample_idx]
+                    dist += _euclidean_distance(target_kdma['value'], sample)
+                if dist < min_dist:
+                    min_dist = dist
+                    best_sample_index = sample_idx
+            return selected_choice, probs, best_sample_index
```

Comment:
This may be me not being capable of math/logic this late in the day, but is this (and alternatives) not doing essentially the same thing as `_select_min_dist_choice` to find the index? Should that function just return an index? I'm also not sure if this logic works for misaligned

I'm also not sure I'm sold on the variable number of returns, but I feel like @dmjoy would have thoughts on that

Reply:
It is super similar but different. At first we are trying to select the **choice** with the min distance by aggregating over the samples. Once the choice is selected, we have to find which **sample** had the min distance for that choice because that is the "best" sample and we want to use it's reasoning as justification. Does that make sense? I don't love all the almost identical looking but different logic in the code, but I'm not sure how else to do it.

Reply:
Ah yes, gotcha

Reply:
I don't have strong feelings about the variable number of returns. If that is something we typically avoid we can make it always return the best_sample_index. I just made it an option because it doesn't make much sense in the context of the oracle which doesn't have samples so will always return 0

Reply:
I don't think this will work for misaligned... Maybe there is a way to reuse the _select_min_dist logic. I'll work on it

Reply:
If we know what the distance of the selected choice was, we just need to choose the one closest to *that* if that makes sense

Reply:
If I'm following, you're suggesting selecting the sample with distance closest to the distance of the choice (aggregate over samples), which would give us sort of an average sample. I think we do want the best sample here - or closest to the target

Reply:
I think that would work, I'll try it

Reply:
I've moved it, I think it is cleaner as a separate function. Let me know what you think https://github.com/ITM-Kitware/align-system/pull/98/commits/9f5b2acc462e64cbfc3e4e90b8a6c77541d837b3